modules/transformer-batch/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/batch/badrows/KinesisSink.scala

```diff
@@ -0,0 +1,248 @@
+    new KinesisSink(writeDataToKinesis, config)
+  }
+
+  final case class TryBatchResult(
+    nextBatchAttempt: Vector[PutRecordsRequestEntry],
+    hadSuccess: Boolean,
+    wasThrottled: Boolean,
+    exampleInternalError: Option[String]
+  ) {
+    // Only retry the exact same again if no record was successfully inserted, and all the errors
+    // were not throughput exceeded exceptions
+    def shouldRetrySameBatch: Boolean =
+      !hadSuccess && !wasThrottled
+  }
+
+  object TryBatchResult {
+
+    implicit def tryBatchResultMonoid: Monoid[TryBatchResult] =
+      new Monoid[TryBatchResult] {
+        override val empty: TryBatchResult = TryBatchResult(Vector.empty, hadSuccess = false, wasThrottled = false, None)
+        override def combine(x: TryBatchResult, y: TryBatchResult): TryBatchResult =
+          TryBatchResult(
+            x.nextBatchAttempt ++ y.nextBatchAttempt,
+            x.hadSuccess || y.hadSuccess,
+            x.wasThrottled || y.wasThrottled,
+            x.exampleInternalError.orElse(y.exampleInternalError)
+          )
+      }
+
+    def build(records: List[PutRecordsRequestEntry], result: PutRecordsResult): TryBatchResult =
+      if (result.getFailedRecordCount.toInt =!= 0)
+        records
+          .zip(result.getRecords.asScala)
+          .foldMap { case (orig, recordResult) =>
+            Option(recordResult.getErrorCode) match {
+              case None =>
+                TryBatchResult(Vector.empty, hadSuccess = true, wasThrottled = false, None)
+              case Some("ProvisionedThroughputExceededException") =>
+                TryBatchResult(Vector(orig), hadSuccess = false, wasThrottled = true, None)
+              case Some(_) =>
+                TryBatchResult(Vector(orig), hadSuccess = false, wasThrottled = false, Option(recordResult.getErrorMessage))
+            }
+          }
+      else
+        TryBatchResult(Vector.empty, hadSuccess = true, wasThrottled = false, None)
+  }
+
+  final case class KeyedData(key: String, content: Array[Byte]) {
+    val size = content.length + key.getBytes(UTF_8).length
+  }
+
+  final case class Batch(
+    partitionIndex: Int,
+    size: Int,
+    recordsCount: Int,
+    keyedData: List[KeyedData]
+  ) {
+    def addData(data: KeyedData): Batch =
+      Batch(this.partitionIndex, size + data.size, recordsCount + 1, data :: keyedData)
+
+    def asString: String = s"Partition: $partitionIndex, initial size in bytes: $size, records count: $recordsCount"
+  }
+
+  object Batch {
+    def empty(partition: Int) = Batch(partition, size = 0, recordsCount = 0, keyedData = List.empty)
+  }
+
+  object Retries {
+
+    def fullJitter(config: BackoffPolicy): RetryPolicy[Try] =
+      capBackoffAndRetries(config, RetryPolicies.fullJitter[Try](config.minBackoff))
+
+    def fibonacci(config: BackoffPolicy): RetryPolicy[Try] =
+      capBackoffAndRetries(config, RetryPolicies.fibonacciBackoff[Try](config.minBackoff))
```

Comment:
So here in batch transformer I'm relaying on implementation from enrich https://github.com/snowplow/enrich/issues/697

One policy for throttling, the other for some unexpected internal failures

Reply:
I remember that we need 2 policies because throttled should be retried forever and with max 1 second between retries, but I can't remember why fibonacci for them. Can you help @istreeter please?

Reply:
> Fibonacci because it's more aggressive the exponential.

That's what I was assuming, great. Thank you!